lock/src/main/java/org/infinispan/lock/impl/lock/ClusteredLockImpl.java

```diff
@@ -221,34 +255,101 @@ public void entryRemoved(CacheEntryRemovedEvent event) {
 
-         // TODO: Split brain and leaving nodes, handle this better
-         if(newMembers.size() > 1 || (oldMembers.size() == 2 && newMembers.size() == 1)) {
-            // There has to be at least 2 members
-            // FIXME: Reentrant locks, can the node rejoin and reacquire the lock before the release actually happens ?? Use the requestId
-            oldMembers.stream().filter(a -> !newMembers.contains(a)).forEach(notPresent -> {
-               try {
-                  if (clusteredLockManager.isDefined(name)) {
-                     clusteredLockManager.execute(() -> unlock(null, notPresent)
-                           .exceptionally(ex -> {
-                              log.error("Unlock failed wrong", ex);
-                              return null;
-                           }));
-                  }
-               } catch (AvailabilityException ex) {
-                  log.error("Unable to release due to cluster change", ex);
-               }
-            });
+         Set<Object> leavingNodes = oldMembers.stream().filter(a -> !newMembers.contains(a)).collect(Collectors.toSet());
+
+         if (leavingNodes.isEmpty()) {
+            log.tracef("LOCK[%s] Nothing to do, all nodes are present %s", getName(), originator);
+            return;
+         }
+
+         if (leavingNodes.size() >= newMembers.size()) {
+            log.tracef("LOCK[%s] Nothing to do, we are on a minority partition notification on %s", getName(), originator);
+            return;
+         }
+
+         //TODO : Not sure this can happen
+//         if (leavingNodes.contains(originator)) {
+//            log.tracef("LOCK[%s] %s is a leaving node. Empty pending requests ", getName(), originator);
+//            while (!pendingRequests.isEmpty()) {
+//               pendingRequests.poll().handleLockResult(null, log.nodeShutdown());
+//            }
+//            return;
+//         }
+
+
+         if (clusteredLockManager.isDefined(name)) {
+            log.tracef("LOCK[%s] %s launches unlock for each leaving node", getName(), originator);
+            forceUnlockForLeavingMembers(leavingNodes);
          }
       }
+
+      /**
+       * This method forces unlock for each of the Address that is not present in the cluster. We don't know which node
+       * holds the lock, so we force an unlock
+       *
+       * @param possibleOwners
+       */
+      private void forceUnlockForLeavingMembers(Set<Object> possibleOwners) {
+         clusteredLockManager.execute(() -> {
+                  log.tracef("LOCK[%s] Call force unlock for %s from %s ", getName(), possibleOwners, originator);
+                  viewChangeUnlockHappening.set(true);
+                  unlock(null, possibleOwners)
+                        .whenComplete((unlockResult, ex) -> {
+                           log.tracef("LOCK[%s] Force unlock call completed for %s from %s ", getName(), possibleOwners, originator);
+                           viewChangeUnlockHappening.set(false);
```

Comment:
I am wondering if a boolean is sufficient. If we have 2 nodes leave consecutively, but not at the same time I think we could get into a state where lock acquisitions could occur. If so we could use an AtomicInteger instead and only fire pending requests if it was zero after decrementing.

Reply:
right, nice catch